cpu/sam0_common/usb/usb.c

```diff
@@ -0,0 +1,671 @@
+            USB_DEVICE_EPINTENSET_STALL1);
+}
+
+static bool usb_enable_syncing(void)
+{
+    if (USB->DEVICE.SYNCBUSY.bit.ENABLE) {
+        return true;
+    }
+    return false;
+}
+
+static bool usb_swrst_syncing(void)
+{
+    if (USB->DEVICE.SYNCBUSY.bit.SWRST) {
+        return true;
+    }
+    return false;
+}
+
+static usbdev_ep_t *usbdev_new_ep(usbdev_t *dev, usb_ep_type_t type,
+                                  usb_ep_dir_t dir, size_t buf_len)
+{
+    /* The IP supports all types for all endpoints */
+    (void)dev;
+    usbdev_ep_t *res = NULL;
+    if (type == USB_EP_TYPE_CONTROL) {
+        res = _get_ep(0, dir);
+        res->num = 0;
+    }
+    else {
+        for (unsigned idx = 1; idx < 8; idx++) {
+            usbdev_ep_t *ep = _get_ep(idx, dir);
+            if (ep->type == USB_EP_TYPE_NONE) {
+                res = ep;
+                res->num = idx;
+                break;
+            }
+        }
+    }
+    if (res) {
+        res->dir = dir;
+        if (_usbdev->used + buf_len < SAM_USB_BUF_SPACE) {
+            res->buf = _usbdev->buffer + _usbdev->used;
+            _usbdev->used += buf_len;
+            res->len = buf_len;
+            _ep_address(res);
+            _ep_size(res);
+        }
+        else {
+            return NULL;
+        }
+        res->type = type;
+        res->cb = NULL;
+        res->driver = &driver_ep;
+    }
+    return res;
+}
+
+static inline void poweron(void)
+{
+#if defined(CPU_FAM_SAMD21)
+    PM->AHBMASK.reg |= PM_AHBMASK_USB;
+    PM->APBBMASK.reg |= PM_APBBMASK_USB;
+    GCLK->CLKCTRL.reg = (uint32_t)(GCLK_CLKCTRL_CLKEN |
+                                   GCLK_CLKCTRL_GEN_GCLK0 |
+                                   (GCLK_CLKCTRL_ID(USB_GCLK_ID)));
+#elif defined(CPU_FAM_SAML21)
+    MCLK->AHBMASK.reg |= MCLK_AHBMASK_USB;
+    GCLK->PCHCTRL[USB_GCLK_ID].reg = GCLK_PCHCTRL_CHEN |
+                                     GCLK_PCHCTRL_GEN_GCLK0;
+#endif
+}
+
+static inline void poweroff(void)
+{
+#if defined(CPU_FAM_SAMD21)
+    PM->AHBMASK.reg |= PM_AHBMASK_USB;
+    GCLK->CLKCTRL.reg = (uint32_t)(GCLK_CLKCTRL_CLKEN |
+                                   GCLK_CLKCTRL_GEN_GCLK0 |
+                                   (GCLK_CLKCTRL_ID(USB_GCLK_ID)));
+#elif defined(CPU_FAM_SAML21)
+    MCLK->AHBMASK.reg |= (MCLK_AHBMASK_USB);
+    GCLK->PCHCTRL[USB_GCLK_ID].reg = GCLK_PCHCTRL_CHEN |
+                                     GCLK_PCHCTRL_GEN_GCLK0;
+#endif
+}
+
+void usbdev_init(usbdev_t *dev)
+{
+    /* Only one usb device on this board */
+    _usbdev = (sam0_common_usb_t *)dev;
+    _usbdev->used = 0;
+    /* Set GPIO */
+    gpio_init(GPIO_PIN(PA, 24), GPIO_IN);
+    gpio_init(GPIO_PIN(PA, 25), GPIO_IN);
+    gpio_init_mux(GPIO_PIN(PA, 24), GPIO_MUX_G);
+    gpio_init_mux(GPIO_PIN(PA, 25), GPIO_MUX_G);
+    poweron();
+    /* Reset peripheral */
+    USB->DEVICE.CTRLA.reg |= USB_CTRLA_SWRST;
+    while (usb_swrst_syncing()) {}
+    while (USB->DEVICE.CTRLA.bit.SWRST) {}
```

Comment:
You can remove this line. The datasheet states that both bit will be clearer once reset is done, so the line above should be enough.

Reply:
Duplicate something something...